design/keycloak.x/configuration.md

```diff
@@ -0,0 +1,159 @@
+# Keycloak.X Server Configuration
+
+* **Status**: Draft #1
+
+## How to set configuration
+
+Keycloak.X can be configured in a property file, using command-line arguments or environment variables. All configuration
+properties are available in all different approaches.
+
+Properties (conf/keycloak.properties):
+
+    hostname.frontendUrl=https://mykeycloak
+    
+Environment variables:
+
+    export KC_HOSTNAME_FRONTEND-URL=https://mykeycloak
+    bin/kc.sh
+
+Command-line arguments:
+
+    bin/kc.sh --hostname-frontendUrl=https://mykeycloak
+    
+When evaluating configuration if a property is defined in several places the order of resolution is first in the list of:
+command-line arguments, environment variables, properties.  
+
+The format used in properties will be used throughout the documentation.
+
+The keys used above follows a strict rewrite rule, where the key in properties is the name of the property. Property 
+names should follow the following rules:
+
+* Use camel-case (`propertyName`)
+* Use dot (`.`) for separation
+* All options must be in a category (`<categoryName>.<propertyName>`)
+* Multiple levels of categories are permitted (`<categoryName>.<subCategoryName>.<propertyName>`)
+
+Rules for converting property keys to different formats follow these rules:
+
+* Environment variables: 
+  - Prefix with `KC_`
+  - Characters converted to uppercase
+  - Dot (`.`) replaced with underscore (`_`)
+  - Use hyphen (`-`) for camel-case names (`PROPERTY-NAME`)
+* Command-line arguments:
+  - Dot (`.`) converted to hyphen (`-`)
+  
+## SPIs and Provider Configuration
+
+Properties for providers follow the following format:
+
+    spi.<spi-name>.<provider-name>.<config-name>
+
+For example:
+
+    spi.hostname.default.frontendUrl=https://mykeycloak
```

Comment:
If you have multiple providers for the same spi, how do you specify which one should be used by default?

Reply:
The thinking was `spi.hostname.provider=mycustom`. It does mean you can't have a provider named `provider` as you'd have a name clash, but I think that's unlikely and can't think of a cleaner way to do it.

Reply:
What if we do:

```
# maps to the default
hostname.frontendUrl=https://mykeycloak

# maps to a specific provider
hostname.default.frontendUrl=https://mykeycloak
hostname.custom.frontendUrl=https://mykeycloak
```

The logic to map to the correct provider would be based on checking whether or not the part after the SPI name can reference a provider.

Reply:
Are you saying to remove `spi` part? That's there to prevent collisions, as we could end up having SPIs with a name that conflicts with other "top-level" categories.

I don't understand what you mean about "The logic to map to the correct provider would be based on checking whether or not the part after the SPI name can reference a provider"

Reply:
Not removing the `spi` part. But just avoiding `spi.hostname.provider` to tell which provider to configure.

Reply:
I'm not really following what you are proposing then ;)

Reply:
how about:

# Provider configuration
```
hostname.default.frontendUrl=https://mykeycloak
hostname.custom.frontendUrl=https://mykeycloak
hostname.custom.foo=bar
```

# Provider selection

## If you want the default, use:
```
spi.hostname=default
```

## If you want the custom, use: (also via -Dspi.hostname=custom)
```
spi.hostname=custom
```

Reply:
Looks better.

My proposal was based on the assumption that you wouldn't need to set the provider id/alias as you only have a single one defined in your configuration. Like for `hostname', you would have either `default` or `custom`, right? So, based on this assumption we wouldn't need to set the provider id/alias. Not sure if this always makes sense.

Reply:
I think it needs to be explicit as there could be no configuration for the provider you want to use as default.

Reply:
`spi.hostname=default` is shorter and kinda nicer, but it would be problematic if for some weird reason we wanted to support a structured format like JSON in the future (don't think we would though). Still not 100% sure about it though. Actually thinking it might be better to make it even more explicit like `spi.hostname.defaultProvider=myprovider`. That's kinda fuggly though.

Reply:
I'm sticking with `spi.hostname.provider` as it is not that commonly set, it's more clear and avoids any issues with conflicts in structured config formats if we want to include that in the future.